consensus/enclave/impl/src/fee_map.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A helper object for maintaining a map of token id -> minimum fee.
+
+use alloc::{collections::BTreeMap, string::String};
+use mc_crypto_digestible::{DigestTranscript, Digestible, MerlinTranscript};
+use mc_sgx_compat::sync::Mutex;
+use mc_transaction_core::{constants::MINIMUM_FEE, tx::TokenId};
+
+/// State managed by `FeeMap`.
+struct FeeMapInner {
+    /// The actual map of token_id to fee.
+    /// Since we hash this map, it is important to use a BTreeMap as it
+    /// guarantees iterating over the map is in sorted and predictable
+    /// order.
+    pub map: BTreeMap<TokenId, u64>,
+
+    /// Cached digest value, formatted as a string.
+    /// (Suitable for appending to responder id)
+    pub cached_digest: String,
+}
+
+impl Default for FeeMapInner {
+    fn default() -> Self {
+        let mut map = BTreeMap::new();
+        map.insert(TokenId::MOB, MINIMUM_FEE);
+
+        let cached_digest = calc_digest_for_map(&map);
+
+        Self { map, cached_digest }
+    }
+}
+
+/// A thread-safe object that contains a map of fee value by token id.
+pub struct FeeMap {
+    inner: Mutex<FeeMapInner>,
+}
+
+impl Default for FeeMap {
+    fn default() -> Self {
+        Self {
+            inner: Mutex::new(FeeMapInner::default()),
+        }
+    }
+}
+
+impl From<BTreeMap<TokenId, u64>> for FeeMap {
+    fn from(map: BTreeMap<TokenId, u64>) -> Self {
+        let cached_digest = calc_digest_for_map(&map);
+
+        Self {
+            inner: Mutex::new(FeeMapInner { map, cached_digest }),
+        }
+    }
+}
+
+impl FeeMap {
+    /// Get the digest of the fee map as a hex-encoded string
+    pub fn get_digest_str(&self) -> String {
+        let inner = self.inner.lock().unwrap();
+        inner.cached_digest.clone()
+    }
+
+    /// Get the fee for a given token id, or None if no fee is set for that
+    /// token.
+    pub fn get_fee_for_token(&self, token_id: &TokenId) -> Option<u64> {
+        let inner = self.inner.lock().unwrap();
+        inner.map.get(token_id).cloned()
+    }
+
+    /// Update the fee map with a new one if provided, or reset it to the
+    /// default.
+    pub fn update_or_default(&self, minimum_fees: Option<BTreeMap<TokenId, u64>>) {
```

Comment:
Why do we want this behavior vs a separate `update` and `default` method? (This is just curiosity - not suggesting a change)

Reply:
It makes the enclave code slightly simpler, but it can also be changed to `update()` and `reset_to_default()` or something like that. I don't think there's much of a difference since currently it is only used in one place in the enclave init code.